internal/pipeline/executor.go

```diff
@@ -0,0 +1,271 @@
+		default:
+			err = fmt.Errorf("invalid pkg state %v detected", state)
+			return
+		}
+	}
+	// mark the pkg in hydrating
+	hctx.hydrationStates[p.path] = Hydrating
+	// add to the graph
+	hctx.pkgs[p.path] = p
+
+	// read the pkg, the pipeline, dependencies etc.
+	err = readPkg(p)
+	if err != nil {
+		err = fmt.Errorf("failed to read the pkg %w", err)
+		return
+	}
+
+	// input resources for current pkg's hydration
+	var input []*yaml.RNode
+	includeLocalResources := false
+	// here we explore the p.sources to determine sub packages etc.
+	for _, s := range expand(p.pipeline.Sources, p.path) {
+		// TODO(droot): sync with pipeline's defaults once Donny's PR is merged
+		if s == "." {
+			// current pkg
+			// will be hydrated as we drop of this loop
+			includeLocalResources = true
+		} else {
+			// TODO(droot): recursive hydration is broken for now, needs more work
+			var depResources []*yaml.RNode
+			depPkg := &pkg{
+				path: s,
+			}
+			depResources, err = hydrate(depPkg, hctx)
+			if err != nil {
+				err = fmt.Errorf("failed to hydrated sub pkg %s %w", s, err)
+				return
+			}
+			input = append(input, depResources...)
+		}
+	}
+
+	if includeLocalResources {
+		var localResources []*yaml.RNode
+		// TODO(droot): figure out how to avoid reading sub packages here ?
+		pkgReader := &kio.LocalPackageReader{PackagePath: p.path, MatchFilesGlob: kio.MatchAll}
+		localResources, err = pkgReader.Read()
+		if err != nil {
+			err = fmt.Errorf("failed to read resources for pkg %s %w", p.path, err)
+			return
+		}
+		// TODO(droot): this will be the place where we filter kpt resources
+		localResources = filter(localResources)
+		input = append(input, localResources...)
+	}
+
+	output := &kio.PackageBuffer{}
+	// TODO(droot): parameterize the sink-mode (hctx.sinkMode) to determine whether to
+	// write it in-place or not. Currently in-place.
+	pkgWriter := &kio.LocalPackageWriter{PackagePath: p.path}
+	pipe := kio.Pipeline{
+		Inputs: []kio.Reader{
+			&kio.PackageBuffer{Nodes: input},
+		},
+		Filters: fnFilters(p.pipeline, logger),   // we will gather filters from the pipeline
+		Outputs: []kio.Writer{pkgWriter, output}, // here may be we don't want to write to the fs yet
+	}
+	err = pipe.Execute()
+	if err != nil {
+		err = fmt.Errorf("failed to run pipeline for pkg %s %w", p.path, err)
+		return
+	}
+
+	// pkg is hydrated
+	resources = output.Nodes
+	// mark the pkg as wet and update the resources
+	hctx.hydrationStates[p.path] = Wet
+	p.resources = resources
+	return
+}
+
+func readPkg(p *pkg) error {
+	// Read the pipeline for the current package else assume default
+	// TODO(droot): integrate with Donny's code once that is merged
+	// and read an actual pipeline
+	p.pipeline = New()
+
+	// adjust the pipeline.sources to point to
+	return nil
+}
+
+// expands takes a list of sources (./*, ./, ...) and current pkg path
+// and returns list of package paths that this pkg depends on.
+func expand(sources []string, pkgPath string) []string {
```

Comment:
p.normalizedSources()?

Regardless of where validation logic runs, this needs to guarantee that source are normalized/clean/expanded. This logic will need to be well unit-tested and it will grow in complexity over time

Reply:
sgtm